lib/strategies/webapp-strategy.js

```diff
@@ -457,27 +458,30 @@ function retrieveTokens(formData, strategy) {
 				if (tokens.length < 2 || !tokens[0] || !tokens[1]) {
 					throw new Error("Invalid access/id token");
 				}
-
-				if (!TokenUtil.validateIssAndAud(tokens[0], serviceConfig)) {
-					throw new Error("Access token validation failed");
-				}
-
-				if (!TokenUtil.validateIssAndAud(tokens[1], serviceConfig)) {
-					throw new Error("Id token validation failed");
-				}
-				appIdAuthorizationContext.accessToken = accessTokenString;
-				appIdAuthorizationContext.accessTokenPayload = tokens[0];
-				appIdAuthorizationContext.identityToken = identityTokenString;
-				appIdAuthorizationContext.identityTokenPayload = tokens[1];
-				logger.debug("authentication success");
-				deferred.resolve(appIdAuthorizationContext);
+				const [accessToken, idToken] = tokens;
+
+				return TokenUtil.validateIssAndAud(accessToken, serviceConfig)
+					.then(()=>TokenUtil.validateIssAndAud(idToken, serviceConfig))
+					.then(() => {
+						appIdAuthorizationContext.accessToken = accessTokenString;
+						appIdAuthorizationContext.accessTokenPayload = tokens[0];
+						appIdAuthorizationContext.identityToken = identityTokenString;
+						appIdAuthorizationContext.identityTokenPayload = tokens[1];
+						logger.debug("authentication success");
+						resolve(appIdAuthorizationContext);
+					})
+					.catch(err => {
+						logger.error(err);
+						throw new Error("token validation failed")
+					})
 			}).catch(function (error) {
 				logger.debug("Authentication failed : " + error.message);
```

Comment:
i'd suggest to log the entire error object instead only the message

Reply:
I don't know if the end user cares about our stack trace, he wants to know what happned + it would be changing the current behavior.

Reply:
Adding more logs isn't changing of behavior. Let's add it.
What do you mean by "end user"? I suggest only to `log` the error, means that Dave would be able to see the error (and paste it in the ticket he opens for us.)